### PR TITLE
EN-3698 : start refactoring termui console

### PR DESCRIPTION
**Pull request by miiu96**
termui refactoring

### CODE REVIEW

File: statusHandler/presenterStatusHandler.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+func (psh *PresenterStatusHandler) GetNetworkSentBps() uint64 {
+	return psh.getFromCacheAsUint64(core.MetricNetworkSentBps)
+}
+
+// GetNetworkSentBpsPeak will return sent bytes per seconds peak
+func (psh *PresenterStatusHandler) GetNetworkSentBpsPeak() uint64 {
+	return psh.getFromCacheAsUint64(core.MetricNetworkSentBpsPeak)
+}
+
+func (psh *PresenterStatusHandler) getFromCacheAsUint64(metric string) uint64 {
```
Comment (iulianpascalau) [inline statusHandler/presenterStatusHandler.go]: :+1: for using wrapping funcs

File: statusHandler/presenterStatusHandler.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+)
+
+//maxLogLines is used to specify how many lines of logs need to store in slice
+var maxLogLines = 100
+
+// PresenterStatusHandler will be used when an AppStatusHandler is required, but another one isn't necessary or available
+type PresenterStatusHandler struct {
```
Comment (iulianpascalau) [inline statusHandler/presenterStatusHandler.go]: Might spawn this struct over multiple files (blockInfoGetters.go, chainInfoGetters.go...)

Comment (iulianpascalau) [inline statusHandler/presenterStatusHandler.go]: Struct's code coverage is below 80%

File: statusHandler/interface.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package statusHandler
```
Comment (iulianpascalau) [inline statusHandler/interface.go]: might move this file in statusHandler/view package

File: statusHandler/presenterStatusHandler.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
+package statusHandler
```
Comment (iulianpascalau) [inline statusHandler/presenterStatusHandler.go]: might move this file in statusHandler/presenter package

File: statusHandler/termuic/termuiConsole.go [hunk 5]
```diff
@@ -3,64 +3,33 @@ package termuic
 import (
```
Comment (iulianpascalau) [inline statusHandler/termuic/termuiConsole.go]: might move this file in statusHandler/views/termuic package

File: statusHandler/termuic/termuiRenders/widgetsRender.go [hunk 6]
```diff
@@ -3,10 +3,10 @@ package termuiRenders
 import (
```
Comment (iulianpascalau) [inline statusHandler/termuic/termuiRenders/widgetsRender.go]: might move this file in statusHandler/views/termuic/termuiRenders package

File: statusHandler/termuic/termuiRenders/widgetsRender.go [hunk 7]
```diff
@@ -27,14 +27,14 @@ type WidgetsRender struct {
...
 	networkSent *widgets.Gauge
 
-	termuiConsoleMetrics *sync.Map
+	presenter statusHandler.PresenterInterface
 }
 
 //NewWidgetsRender method will create new WidgetsRender that display termui console
-func NewWidgetsRender(metricData *sync.Map, grid *DrawableContainer) *WidgetsRender {
+func NewWidgetsRender(presenter statusHandler.PresenterInterface, grid *DrawableContainer) *WidgetsRender {
 	self := &WidgetsRender{
```
Comment (iulianpascalau) [inline statusHandler/termuic/termuiRenders/widgetsRender.go]: should add checks for nil and should add tests for those checks

File: statusHandler/presenterStatusHandler.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+	"strings"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+)
+
+//maxLogLines is used to specify how many lines of logs need to store in slice
+var maxLogLines = 100
+
+// PresenterStatusHandler will be used when an AppStatusHandler is required, but another one isn't necessary or available
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/presenterStatusHandler.go]: Done.

File: statusHandler/presenterStatusHandler.go [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+}
+
+// GetConsensusRoundState will return consensus round state
+func (psh *PresenterStatusHandler) GetConsensusRoundState() string {
+	return psh.getFromCacheAsString(core.MetricConsensusRoundState)
+}
+
+// GetLogLines will return log lines that's need to be displayed
+func (psh *PresenterStatusHandler) GetLogLines() []string {
+	return psh.logLines
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/presenterStatusHandler.go]: Done.

File: statusHandler/presenterStatusHandler.go (hunk 2 above)
Comment (miiu96) [reply statusHandler/presenterStatusHandler.go]: Done.

File: statusHandler/interface.go (hunk 3 above)
Comment (miiu96) [reply statusHandler/interface.go]: Done.

File: statusHandler/presenterStatusHandler.go (hunk 4 above)
Comment (miiu96) [reply statusHandler/presenterStatusHandler.go]: Done.

File: statusHandler/termuic/termuiConsole.go (hunk 5 above)
Comment (miiu96) [reply statusHandler/termuic/termuiConsole.go]: Done.

File: statusHandler/termuic/termuiRenders/widgetsRender.go (hunk 6 above)
Comment (miiu96) [reply statusHandler/termuic/termuiRenders/widgetsRender.go]: Done.

File: statusHandler/termuic/termuiRenders/widgetsRender.go (hunk 7 above)
Comment (miiu96) [reply statusHandler/termuic/termuiRenders/widgetsRender.go]: Checks for nil was added for constructor parameters and if an parameter is nil an error will be returned.

File: cmd/node/main.go [hunk 10]
```diff
@@ -40,6 +40,7 @@ import (
 	"github.com/ElrondNetwork/elrond-go/process/smartContract/hooks"
 	"github.com/ElrondNetwork/elrond-go/sharding"
 	"github.com/ElrondNetwork/elrond-go/statusHandler"
+	"github.com/ElrondNetwork/elrond-go/statusHandler/termuic"
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/main.go]: Done. Now its a ''factory' that creates view and a factory for presenterStatusHandler.

File: cmd/node/main.go [hunk 11]
```diff
@@ -498,21 +499,24 @@ func startNode(ctx *cli.Context, log *logger.Logger, version string) error {
 		appStatusHandlers = append(appStatusHandlers, prometheusStatusHandler)
 	}
 
+	presenterStatusHandler := statusHandler.NewPresenterStatusHandler()
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/main.go]: Done.

File: core/statistics/machine/netStatistics.go [hunk 12]
```diff
@@ -18,14 +20,7 @@ type NetStatistics struct {
...
 func (ns *NetStatistics) ComputeStatistics() {
-
-	currentProcess, err := GetCurrentProcess()
-	if err != nil {
-		ns.setZeroStatsAndWait()
-		return
-	}
-
-	nStart, err := currentProcess.NetIOCounters(false)
+	nStart, err := net.IOCounters(false)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/statistics/machine/netStatistics.go]: net.IOCounters(false) return an structure that contains information about network traffic.

File: statusHandler/presenter/blockInfoGetters_test.go [hunk 13]
```diff
@@ -0,0 +1,62 @@
+package presenter
+
+import (
+	"github.com/ElrondNetwork/elrond-go/core"
```
Comment (iulianpascalau) [inline statusHandler/presenter/blockInfoGetters_test.go]: unsorted imports

File: statusHandler/presenter/chainInfoGetters_test.go [hunk 14]
```diff
@@ -0,0 +1,117 @@
+package presenter
+
+import (
+	"github.com/ElrondNetwork/elrond-go/core"
```
Comment (iulianpascalau) [inline statusHandler/presenter/chainInfoGetters_test.go]: unsorted imports

File: statusHandler/presenter/instanceInfoGetters_test.go [hunk 15]
```diff
@@ -0,0 +1,95 @@
+package presenter
+
+import (
+	"github.com/ElrondNetwork/elrond-go/core"
```
Comment (iulianpascalau) [inline statusHandler/presenter/instanceInfoGetters_test.go]: unsorted imports

File: statusHandler/view/termuic/termuiConsole.go [hunk 16]
```diff
@@ -1,13 +1,16 @@
 package termuic
 
 import (
+	"errors"
+	"github.com/ElrondNetwork/elrond-go/statusHandler"
+	"github.com/prometheus/common/log"
```
Comment (iulianpascalau) [inline statusHandler/view/termuic/termuiConsole.go]: should not use this package for logging
Also unsorted imports

File: statusHandler/view/termuic/termuiRenders/widgetsRender.go [hunk 17]
```diff
@@ -2,11 +2,12 @@ package termuiRenders
 
 import (
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/statusHandler"
```
Comment (iulianpascalau) [inline statusHandler/view/termuic/termuiRenders/widgetsRender.go]: unsorted imports

File: cmd/node/factory/structs.go [hunk 18]
```diff
@@ -596,6 +599,35 @@ func (ncv *nullChronologyValidator) IsInterfaceNil() bool {
...
 
+// PresenterStatusHandlerFactory will return an instance of PresenterStatusHandler
+func PresenterStatusHandlerFactory() view.Presenter {
+	presenterStatusHandlerFactory := factoryPresenter.NewPresenterFactory()
+
+	return presenterStatusHandlerFactory.Create()
+}
+
+// ViewsFactory will start an termui console  and will return an object if cannot create and start termuiConsole
+func ViewsFactory(presenter view.Presenter) ([]statusHandler.Viewer, error) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/factory/structs.go]: Done. Function was renamed with CreateViews.

File: cmd/node/factory/structs.go [hunk 19]
```diff
@@ -596,6 +599,35 @@ func (ncv *nullChronologyValidator) IsInterfaceNil() bool {
 	return false
 }
 
+// PresenterStatusHandlerFactory will return an instance of PresenterStatusHandler
+func PresenterStatusHandlerFactory() view.Presenter {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/factory/structs.go]: Done. Function was renamed.

File: cmd/node/main.go [hunk 20]
```diff
@@ -492,31 +492,40 @@ func startNode(ctx *cli.Context, log *logger.Logger, version string) error {
...
+		writer, ok := presenterStatusHandler.(io.Writer)
+		if ok {
+			err = log.ChangePrinterHookWriter(writer)
+			if err != nil {
+				return err
+			}
 		}
 
-		appStatusHandlers = append(appStatusHandlers, presenterStatusHandler)
+		statusHandlerO, ok := presenterStatusHandler.(core.AppStatusHandler)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/main.go]: statusHandler0 was not a good name. Variable was renamed with name "appStatusHandler".

File: statusHandler/factory/viewersFactory.go [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+// Create returns an view slice that will hold all views in the system
+func (wf *viewsFactory) Create() ([]statusHandler.Viewer, error) {
+	viewers := make([]statusHandler.Viewer, 0)
+
+	termuiConsole, err := wf.createTermuiConsole()
+	if err != nil {
+		return nil, err
+	}
+	viewers = append(viewers, termuiConsole)
+	return viewers, nil
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/factory/viewersFactory.go]: Done. Also the name of file was renamed with "viewsFactory.go".

File: statusHandler/presenter/common.go [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+		return 0
+	}
+
+	return valUint64
+}
+
+func (psh *PresenterStatusHandler) getFromCacheAsString(metric string) string {
+	val, ok := psh.presenterMetrics.Load(metric)
+	if !ok {
+		return "[invalid key]"
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/presenter/common.go]: Done.

File: statusHandler/interface.go [hunk 23]
```diff
@@ -1,39 +1,5 @@
...
-	GetTotalMem() uint64
-	GetMemUsedByNode() uint64
-	GetNetworkRecvPercent() uint64
-	GetNetworkRecvBps() uint64
-	GetNetworkRecvBpsPeak() uint64
-	GetNetworkSentPercent() uint64
-	GetNetworkSentBps() uint64
-	GetNetworkSentBpsPeak() uint64
-	GetLogLines() []string
+type Viewer interface {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/interface.go]: Done. Viewer interface was moved.

File: statusHandler/presenter/chainInfoGetters_test.go (hunk 14 above)
Comment (miiu96) [reply statusHandler/presenter/chainInfoGetters_test.go]: Done. Imports was sorted.

File: statusHandler/presenter/instanceInfoGetters_test.go (hunk 15 above)
Comment (miiu96) [reply statusHandler/presenter/instanceInfoGetters_test.go]: Done. Imports was sorted.

File: statusHandler/presenter/blockInfoGetters_test.go (hunk 13 above)
Comment (miiu96) [reply statusHandler/presenter/blockInfoGetters_test.go]: Done.

File: statusHandler/view/termuic/termuiConsole.go (hunk 16 above)
Comment (miiu96) [reply statusHandler/view/termuic/termuiConsole.go]: Done. Logger was changed and imports was sorted.

File: statusHandler/view/termuic/termuiRenders/widgetsRender.go (hunk 17 above)
Comment (miiu96) [reply statusHandler/view/termuic/termuiRenders/widgetsRender.go]: Done.

File: statusHandler/presenter/common.go [hunk 24]
```diff
@@ -0,0 +1,29 @@
+package presenter
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/presenter/common.go]: Done.

File: statusHandler/presenter/presenterStatusHandler.go [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+	psh.presenterMetrics.Store(key, value)
+}
+
+// SetStringValue method - will update the value of a key
+func (psh *PresenterStatusHandler) SetStringValue(key string, value string) {
+	psh.presenterMetrics.Store(key, value)
+}
+
+// Increment - will increment the value of a key
+func (psh *PresenterStatusHandler) Increment(key string) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/presenter/presenterStatusHandler.go]: Done.

File: statusHandler/factory/presenterFactory.go [hunk 26]
```diff
@@ -0,0 +1,23 @@
+package factory
+
+import (
+	"github.com/ElrondNetwork/elrond-go/statusHandler/presenter"
+	"github.com/ElrondNetwork/elrond-go/statusHandler/view"
+)
+
+type presenterFactory struct {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/factory/presenterFactory.go]: Unit tests was added.

File: statusHandler/factory/viewersFactory.go [hunk 27]
```diff
@@ -0,0 +1,44 @@
+package factory
+
+import (
+	"github.com/ElrondNetwork/elrond-go/statusHandler"
+	"github.com/ElrondNetwork/elrond-go/statusHandler/view"
+	"github.com/ElrondNetwork/elrond-go/statusHandler/view/termuic"
+)
+
+type viewsFactory struct {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply statusHandler/factory/viewersFactory.go]: Unit tests was added.

File: cmd/node/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/node/main.go]: ok!

File: cmd/node/main.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/node/main.go]: ok

File: core/statistics/machine/netStatistics.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply core/statistics/machine/netStatistics.go]: ok

**Pull request merged**